lingvo/tasks/asr/ctc_model.py

```diff
@@ -138,62 +140,118 @@ def __init__(self, params):
 
   def ComputePredictions(self, theta, input_batch):
     output_batch = self._FProp(theta, input_batch)
-    return py_utils.RunOnTpuHost(
+    # ctc_greedy_decoder = merge_repeated = True (def)
+    hypotheses = py_utils.RunOnTpuHost(
       tf.nn.ctc_greedy_decoder,
       output_batch.encoder_outputs,
       py_utils.LengthsFromBitMask(
-        tf.squeeze(output_batch.encoder_outputs_padding, 2), 0)
+        tf.squeeze(output_batch.encoder_outputs_padding, 2), 0),
     )
+    return hypotheses
+        # return tf.Print(hypotheses, [hypotheses], "*******ComputePredictions********:")
+
+  def _CalculateWER(self, input_batch, output_batch):
+      (decoded,), neg_sum_logits = tf.nn.ctc_greedy_decoder(
+          output_batch.encoder_outputs,
+          py_utils.LengthsFromBitMask(
+              tf.squeeze(output_batch.encoder_outputs_padding, 2), 0
+          ),
+      )
+
+      dec = tf.sparse_to_dense(
+          decoded.indices, decoded.dense_shape, decoded.values, default_value=73
+      )
+
+      hyp_str = self.input_generator.IdsToStrings(
+          tf.cast(dec, tf.int32), tf.shape(dec)
```

Comment:
tf.shape() isn't right here. The "shapes" as the second argument to IdsToStrings is the length of each sequence in the batch, not the shape of the tensor (which is essentially length along each dimension). We need to convert the indices array into lengths. Probably a reduce-by-key (where the reduce is the max operation) would do it. I'm trying to figure this out.

Reply:
Either reduce-by-key or tf.scatter_nd.